File: clients/cpp/model/DebugData.cpp
#include "DebugData.hpp"
#include <stdexcept>

namespace model {

Circle::Circle(model::Vec2Double pos, double radius) : pos(pos), radius(radius) { }

// Read Circle from input stream
Circle Circle::readFrom(InputStream& stream) {
    model::Vec2Double pos = model::Vec2Double::readFrom(stream);
    double radius = stream.readDouble();
    return Circle(pos, radius);
}

// Write Circle to output stream
void Circle::writeTo(OutputStream& stream) const {
    pos.writeTo(stream);
    stream.write(radius);
}

// Get string representation of Circle
std::string Circle::toString() const {
    std::stringstream ss;
    ss << "Circle { ";
    ss << "pos: ";
    ss << pos.toString();
    ss << ", ";
    ss << "radius: ";
    ss << radius;
    ss << " }";
    return ss.str();
}

    
// Read DebugData from input stream
DebugData readDebugData(InputStream& stream) {
    switch (stream.readInt()) {
    case 0:
        return Circle::readFrom(stream);
    default:
        throw std::runtime_error("Unexpected tag value");
    }
}

// Write DebugData to output stream
void writeDebugData(const DebugData& value, OutputStream& stream) {
    std::visit([&](auto& arg) {
        using T = std::decay_t<decltype(arg)>;
        if constexpr (std::is_same_v<T, Circle>) {
            stream.write((int) 0);
        }
        arg.writeTo(stream);
    }, value);
}

// Get string representation of DebugData
std::string debugDataToString(const DebugData& value) {
    return std::visit([](auto& arg) {
        return arg.toString();
    }, value);
}


}

File: clients/php/Model/DebugData.php
<?php

namespace Model {
    require_once 'Model/Vec2Double.php';
    require_once 'Stream.php';

    /**
     * TODO - Document
     */
    abstract class DebugData
    {
        /**
         * Write DebugData to output stream
         */
        abstract function writeTo(\OutputStream $stream): void;

        /**
         * Read DebugData from input stream
         */
        static function readFrom(\InputStream $stream): DebugData
        {
            $tag = $stream->readInt32();
            if ($tag == \Model\DebugData\Circle::TAG) {
                return \Model\DebugData\Circle::readFrom($stream);
            }
            throw new Exception('Unexpected tag value');
        }
    }
}

namespace Model\DebugData {
    /**
     * TODO - Document
     */
    class Circle extends \Model\DebugData
    {
        const TAG = 0;
    
        /**
         * TODO - Document
         */
        public \Model\Vec2Double $pos;
        /**
         * TODO - Document
         */
        public float $radius;
    
        function __construct(\Model\Vec2Double $pos, float $radius)
        {
            $this->pos = $pos;
            $this->radius = $radius;
        }
    
        /**
         * Read Circle from input stream
         */
        public static function readFrom(\InputStream $stream): Circle
        {
            $pos = \Model\Vec2Double::readFrom($stream);
            $radius = $stream->readDouble();
            return new Circle($pos, $radius);
        }
        
        /**
         * Write Circle to output stream
         */
        public function writeTo(\OutputStream $stream): void
        {
            $stream->writeInt32(Circle::TAG);
            $this->pos->writeTo($stream);
            $stream->writeDouble($this->radius);
        }
    }
}